### PR TITLE
Adding documentation to query.py

### PR DESCRIPTION
**Pull request by mattagape**
I have pretty much got as far as I can with documenting `deviser/generator/util/query.py`. 

Ideally each function will have a docstring with one or two summary lines, then a blank line, then lists of parameters/what's returned, then a blank line, then an example.

I have documented as many of these as possible, but for quite a number I'm really not sure what's going on. This is especially true if one or more of the attributes/returned are undocumented, and/or the presence of a TODO.

But please check all my explanations. Something may look innocuously correct, but I may well have misunderstood what's going on!

Some of the functions I was unable to find client code for, in the deviser/generator code base; might they be used elsewhere?

Thanks

### CODE REVIEW

File: generator/util/query.py [hunk 1]
```diff
@@ -70,17 +97,43 @@ def get_sid_refs_for_class(working_class):
...
+    (i.e. 'type' is 'element', 'lo_element' or 'inline_lo_element')
+
+    :param attributes: the attributes to check
+    :return: True if at least one type match
+
+    e.g.
+    <attribute name="boundaryMin" required="true" type="element"
+     element="Boundary" abstract="false"/>
+
+     TODO is this a correct example of such a node? Or could it be a
```
Comment (skeating) [inline generator/util/query.py]: It could be either

File: generator/util/query.py [hunk 2]
```diff
@@ -121,7 +228,15 @@ def is_inline_child(class_object):
             inline_parents.append(parent)
     return inline_parents
 
+
 def get_inline_parents(class_object):
+    """
```
Comment (skeating) [inline generator/util/query.py:250]: Yep, this and the above need some sorting out issue #35

File: generator/util/query.py [hunk 3]
```diff
@@ -133,16 +248,27 @@ def get_inline_parents(class_object):
...
 def get_parent_class(class_object):
+    """
+    Return the parent class of this class.
+    if it is not given we assume it is a child of a plugin
+    and get the base of the plugin.
+
+    :param class_object: the object representing the class.
+    :return: parent class, if found.
+
+    TODO add example here
```
Comment (skeating) [inline generator/util/query.py]: added with push to branch

File: generator/util/query.py [hunk 4]
```diff
@@ -151,40 +277,58 @@ def get_parent_class(class_object):
...
-                        break
-            if found:
-                break
+                        return base
+                        # parent = base
+                        # found = True
+                        # break
+            # if found:
+            #    break
+
```
Comment (skeating) [inline generator/util/query.py:319]: Added an issue to look at this.

File: generator/util/query.py [hunk 5]
```diff
@@ -151,40 +277,58 @@ def get_parent_class(class_object):
...
 
-# basic function that checks whether the given element is instantiated
 def is_instantiated(element):
+    """
+    Basic function that checks whether the given element is instantiated.
+
+    :param element: the element object to check.
+    :returns: True if instantiated
+
+    TODO add description of line with string equality check
```
Comment (skeating) [inline generator/util/query.py:356]: Not sure what this is asking for ?

File: generator/util/query.py [hunk 6]
```diff
@@ -282,32 +457,69 @@ def get_unique_attributes(full_attributes):
...
+    :return: the element that matches, or None if no match.
+    """
     element = None
     if not list_elements:
         return element
     for existing in list_elements:
         if existing[name] == match_name:
-            element = existing
-    return element
+            element = existing  # TODO could this match multiple times? Would it matter? Why not return now?
```
Comment (skeating) [inline generator/util/query.py]: Agreed - just return when we match

File: generator/util/query.py [hunk 7]
```diff
@@ -282,32 +457,69 @@ def get_unique_attributes(full_attributes):
...
 def has_other_packages(attributes):
+    """
+    Check to see if a list of attributes have other packages.
+
+    :param attributes: the list to check
+    :return: Tuple with True if any of the attributes have another package,
+             plus a list of packages (empty if False).
+
+
+    TODO add example here.
```
Comment (skeating) [inline generator/util/query.py]: Example added and pushed - please check it makes sense and formatting is consistent

File: generator/util/query.py [hunk 8]
```diff
@@ -317,22 +529,44 @@ def has_other_packages(attributes):
     return [has_pack, list_pkgs]
 
 
-# return True is the attribute is saved as a string
 def is_string(attribute):
+    """
+    Is this attribute saved as a string?
```
Comment (skeating) [inline generator/util/query.py]: I would say rather
is the attribute of Type string

File: generator/util/query.py [hunk 9]
```diff
@@ -317,22 +529,44 @@ def has_other_packages(attributes):
...
         return True
     return False
 
 
-# return True is the attribute has an isSet member variable
 def has_is_set_member(attribute):
+    """
+    return True if the attribute has an isSet member variable
+
+    TODO need an example - I can't see why this would return True if has isSet
```
Comment (skeating) [inline generator/util/query.py]: adde and pushed an explanation

File: generator/util/query.py [hunk 10]
```diff
@@ -368,22 +616,35 @@ def isV2BaseAttribute(element, attribute):
...
-            if attrib['element'] == unprefixed_name and attrib['type'] != 'SIdRef':
+            if attrib['element'] == unprefixed_name and \
+                    attrib['type'] != 'SIdRef':
                 if not strFunctions.compare_no_case(name, attrib['xml_name']):
                     overwrites = True
     return overwrites
 
 
 def get_static_extension_attribs(num_versions, lv_info):
+    """
```
Comment (skeating) [inline generator/util/query.py:723]: added explanation and pushed

File: generator/util/query.py [hunk 11]
```diff
@@ -483,8 +776,13 @@ def get_enum(element, class_name = ''):
         return [value, strvalue]
 
 
-
 def get_default_enum_value(attribute):
+    """
```
Comment (skeating) [inline generator/util/query.py:872]: added please check it makes sense

File: generator/util/query.py [hunk 12]
```diff
@@ -496,6 +794,12 @@ def get_default_enum_value(attribute):
 
 
 def get_first_enum_value(attribute):
+    """
```
Comment (skeating) [inline generator/util/query.py:917]: added explanation

File: generator/util/query.py [hunk 13]
```diff
@@ -507,6 +811,12 @@ def get_first_enum_value(attribute):
 
 
 def get_prefix(name):
+    """
```
Comment (skeating) [inline generator/util/query.py]: Ive commented this function out for now

File: generator/util/query.py [hunk 14]
```diff
@@ -523,6 +833,15 @@ def get_prefix(name):
 
 
 def get_typecode_format(name, language):
+    """
+    TODO need an example
```
Comment (skeating) [inline generator/util/query.py]: added

File: generator/util/query.py [hunk 15]
```diff
@@ -533,6 +852,27 @@ def get_typecode_format(name, language):
...
+    <element name="Algebraic" typeCode="SBML_FOO_ALGEBRAIC" ... />
+    <element name="FooRate" typeCode="SBML_FOO_FOORATE" ... />
+    <element name="FooAssignment" typeCode="SBML_FOO_FOOASSIGNMENT" .../>
+
+    a function call like: width = query.get_max_length(elements, 'name')
+
+    will return the length of the longest string in
+    ["Algebraic", "FooRate", "FooAssignment"]
+
+    TODO will this throw an exception if elements other
```
Comment (skeating) [inline generator/util/query.py:1031]: Added query #38

File: generator/util/query.py [hunk 16]
```diff
@@ -546,6 +886,12 @@ def get_max_length(elements, attribute):
 
 
 def get_other_element_children(this_object, element):
+    """
```
Comment (skeating) [inline generator/util/query.py:1049]: please check this one makes sense !!

**Review by skeating — COMMENTED** [review]
@mattgillucl  started this - not got all the way through but you could start to check any explanations I put in make sense

File: generator/util/query.py (hunk 11 above)
Comment (mattagape) [reply generator/util/query.py:872]: 1. Please complete :param: fields
2. typo: facilitate -> facilitates

File: generator/util/query.py (hunk 12 above)
Comment (mattagape) [reply generator/util/query.py:917]: When you say, "Processing of the XML should have added a further value", do you mean that this function is checking the XML generation, and, in your example, it is flagging up an error about the missing value ("OUTPUT_TRANSITION_INVALID")?

File: generator/util/query.py (hunk 16 above)
Comment (mattagape) [reply generator/util/query.py:1049]: Looks good.

File: generator/util/query.py (hunk 12 above)
Comment (skeating) [reply generator/util/query.py:917]: No, I am saying that as part of the parseXML and then further parsing of the results code will automaticallay add an INVALID value to any enumeration

File: generator/util/query.py (hunk 1 above)
Comment (mattagape) [reply generator/util/query.py]: OK, I have removed the TODO comment

File: generator/util/query.py (hunk 5 above)
Comment (mattagape) [reply generator/util/query.py:356]: It's referring to the line `if c['element'] == element['element']:` i.e. what is it checking, and why?

Comment (mattagape) [reply generator/util/query.py:356]: It's referring to the line `if c['element'] == element['element']:` i.e. what is it checking, and why?

File: generator/util/query.py (hunk 6 above)
Comment (mattagape) [reply generator/util/query.py]: OK, I changed the code to reflect this.

File: generator/util/query.py (hunk 12 above)
Comment (mattagape) [reply generator/util/query.py:917]: OK I have added that to the docsrting.

File: generator/util/query.py (hunk 5 above)
Comment (skeating) [reply generator/util/query.py:356]: Added this s issue #43 as it doesnt ever get tested

File: generator/util/query.py [hunk 17]
```diff
@@ -632,8 +642,8 @@ def has_lo_attribute(element, attribute):
     """
     Does the element's listOf class have the attribute specified?
 
-    :param element:
-    :param attribute:
+    :param element: an elelemnt object
```
Comment (mattagape) [inline generator/util/query.py]: typo on 'element'

File: generator/util/query.py [hunk 18]
```diff
@@ -146,13 +147,15 @@ def has_children_not_math(attributes):
 def get_class(name, root_object):
     """
     Return the class with the matching name from the root object.
+
+    :param name: name of th class to find
```
Comment (mattagape) [inline generator/util/query.py]: typo on 'the'

**Review by skeating — APPROVED** [review]

**Pull request merged**